Controllers films users

fzk273 opened the pull request.

Commit:
add User model
add Film model
add UserController
add FilmController
add Validation logic

Commit:
add tests for all classes
refactor REST methods in all controller to public
fix isFilmValid function to work properly

Commit:
fix linting

Commit:
fix linting

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,24 @@
 /**
  * Film.
  */
 @Getter
 @Setter
+@ToString
+@EqualsAndHashCode
```
Поставь просто аннотацию `@Data` и она заменит эти все)

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,24 @@
+@EqualsAndHashCode
 public class Film {
+
+    Long id;
+    String name;
+    String description;
+    LocalDate releaseDate;
+    Integer duration;
```
Не забывай про инкапсуляцию и модификаторы доступа

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,72 @@
+    private static final Logger log = LoggerFactory.getLogger(FilmController.class);
+    private final HashMap<Long, Film> films = new HashMap<>();
+
+
```
Лишний отступ

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,72 @@
+            log.warn("film name is empty");
+            throw new ValidationException("film name is empty");
+        }
+        if (film.getDescription().length() > 200) {
+            log.warn("film description is to long");
+            throw new ValidationException("film description is to long");
+        }
+        if (film.getReleaseDate().isBefore(LocalDate.parse("28-12-1895", DateTimeFormatter.ofPattern("dd-MM-yyyy")))) {
+            log.warn("release date is not valid");
+            throw new ValidationException("release date is not valid");
+        }
+        if (film.getDuration() <= 0) {
+            log.warn("Duration cant be negative");
+            throw new ValidationException("Duration cant be negative");
+        }
```
Ссылочные типы могут быть null, не забывай это проверять тоже, иначе легко можем поймать NPE
Значения для даты выпуска первого фильма и макс длины описания лучше вынести в константы. Формат даты тоже

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,72 @@
+            log.warn("there is no film with id: {}", film.getId());
+            throw new ValidationException("There is no such film");
+        }
+        if (isFilmValid(film)) {
```
Для обновления не совсем подойдет такая же валидация как и для создания, обязательным является только поле айди обновляемой сущности, остальные поля опциональны и их нужно проверить и обновить, только если пришли ненулевые значения в запросе

Валидацию лучше разобраться и реализовать сразу через аннотации, есть неплохая статейка на этот счет, советую ознакомиться
https://struchkov.dev/blog/ru/spring-boot-validation/#

fzk273 replied:
до конца так и не понял как правильно сделать валидацию даты в апдейте.

Ivanov-s-tmn replied:
Оставил комментарий

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,72 @@
 @RestController
+@RequestMapping("/films")
 public class FilmController {
+    private static final Logger log = LoggerFactory.getLogger(FilmController.class);
```
Логгер можно создать аннотацией над классом `@Slf4j` от lombok. А обращаться к нему можно через log

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/test/java/ru/yandex/practicum/filmorate/FilmControllerTests.java:
```diff
@@ -0,0 +1,92 @@
+    public void filmCreationValidatorIsWorking() {
+        film.setReleaseDate(LocalDate.of(1894, 12, 12));
+        Assertions.assertThrows(ValidationException.class, () -> filmController.create(film));
+        film.setReleaseDate(LocalDate.of(2000, 12, 12));
+
+        film.setName(null);
+        Assertions.assertThrows(ValidationException.class, () -> filmController.create(film));
+        film.setName("borat");
+
+        film.setDescription("c".repeat(202));
+        Assertions.assertThrows(ValidationException.class, () -> filmController.create(film));
+        film.setDescription("desc");
+
+        film.setDuration(-1);
+        Assertions.assertThrows(ValidationException.class, () -> filmController.create(film));
```
Лучше разбить тесты на разные, желательно минимизировать количество проверок в тестах, чтобы они были более прогнозируемыми.

Плюс, лучше проверять сообщение исключения валидации. Проверки на всех полях выбрасывают одинаковый тип исключения, но сообщения у них разные и с такой реализацией тестов нет уверенности в том, что исключение выбрасывается там где нам нужно

fzk273 replied:
поправил все тесты

Commit:
refactor tests for UserController
refactor tests for FilmController
add spring boot validation in pom
add validation annotations on User model
add validation annotations on Film model
add validation annotation on FilmController
Add validation annotation on UserController

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -36,32 +37,34 @@ public Film create(@RequestBody Film film) {
         return film;
     }
 
+    //todo тут потенциально возможен баг если заслать невалидную дату (до релиза первого фильма).
+    // как реализовать правильно?
+    // разбить метод изФильмВалид на отдельные методы и тут их вызывать??
```
Реализовать свою аннотацию для валидацию. Или проверять в ручную при обновлении. При ручной проверке тебе не нужна тут аннотация `@Valid`, так как она проверяет многие поля на null, на пустоту, а при обновлении это допустимо. Тебе тут нужно проверить, пришло ли новое значение по полю и если пришло провалидировать и обновить, если нет - ничего не делать.

fzk273 replied:
в итоге решил накостылить с ручной валидацией, на выхах перепишу на свой валидатор.

Commit:
fix update method in FilmController
